src/blockchain_utilities/blockchain_prune.cpp

```diff
@@ -0,0 +1,697 @@
+
+#define MDB_val_set(var, val)   MDB_val var = {sizeof(val), (void *)&val}
+
+namespace po = boost::program_options;
+using namespace epee;
+using namespace cryptonote;
+
+static std::string db_path;
+
+// default to fast:1
+static uint64_t records_per_sync = 128;
+static const size_t slack = 512 * 1024 * 1024;
+
+static std::error_code replace_file(const boost::filesystem::path& replacement_name, const boost::filesystem::path& replaced_name)
+{
+  std::error_code ec = tools::replace_file(replacement_name.string(), replaced_name.string());
+  if (ec)
+    MERROR("Error renaming " << replacement_name << " to " << replaced_name << ": " << ec.message());
+  return ec;
+}
+
+static void open(MDB_env *&env, const boost::filesystem::path &path, uint64_t db_flags, bool readonly)
+{
+  int dbr;
+  int flags = 0;
+
+  if (db_flags & DBF_FAST)
+    flags |= MDB_NOSYNC;
+  if (db_flags & DBF_FASTEST)
+    flags |= MDB_NOSYNC | MDB_WRITEMAP | MDB_MAPASYNC;
+  if (readonly)
+    flags |= MDB_RDONLY;
+
+  dbr = mdb_env_create(&env);
+  if (dbr) throw std::runtime_error("Failed to create LDMB environment: " + std::string(mdb_strerror(dbr)));
+  dbr = mdb_env_set_maxdbs(env, 32);
+  if (dbr) throw std::runtime_error("Failed to set max env dbs: " + std::string(mdb_strerror(dbr)));
+  dbr = mdb_env_open(env, path.string().c_str(), flags, 0664);
+  if (dbr) throw std::runtime_error("Failed to open database file '"
+      + path.string() + "': " + std::string(mdb_strerror(dbr)));
+}
+
+static void close(MDB_env *env)
+{
+  mdb_env_close(env);
+}
+
+static int compare_uint64(const MDB_val *a, const MDB_val *b)
+{
+  const uint64_t va = *(const uint64_t *)a->mv_data;
+  const uint64_t vb = *(const uint64_t *)b->mv_data;
+  return (va < vb) ? -1 : va > vb;
+}
+
+static int compare_hash32(const MDB_val *a, const MDB_val *b)
+{
+  const uint32_t *va = (const uint32_t*) a->mv_data;
+  const uint32_t *vb = (const uint32_t*) b->mv_data;
+  for (int n = 7; n >= 0; n--)
+  {
+    if (va[n] == vb[n])
+      continue;
+    return va[n] < vb[n] ? -1 : 1;
+  }
+
+  return 0;
+}
+
+static int compare_string(const MDB_val *a, const MDB_val *b)
+{
+  const char *va = (const char*) a->mv_data;
+  const char *vb = (const char*) b->mv_data;
+  return strcmp(va, vb);
+}
+
+static void add_size(MDB_env *env, size_t bytes)
+{
+  try
+  {
+    boost::filesystem::path path(db_path);
+    boost::filesystem::space_info si = boost::filesystem::space(path);
+    if(si.available < bytes)
+    {
+      MERROR("!! WARNING: Insufficient free space to extend database !!: " <<
+          (si.available >> 20L) << " MB available, " << (bytes >> 20L) << " MB needed");
+      return;
+    }
+  }
+  catch(...)
+  {
+    // print something but proceed.
+    MWARNING("Unable to query free disk space.");
+  }
+
+  MDB_envinfo mei;
+  mdb_env_info(env, &mei);
+  MDB_stat mst;
+  mdb_env_stat(env, &mst);
+
+  uint64_t new_mapsize = mei.me_mapsize + bytes;
+  new_mapsize += (new_mapsize % mst.ms_psize);
```

Comment:
I assume you mean memcmp. If we do that, then the sorting will be reversed on little/big endian archs, no ?

About the page size logic, do you mean the quantizing ? It probably does it internally, yes. I suppose this can be removed, it just felt better I guess.

Your last sentence got clipped.

Reply:
Two of your comments show with a single reply box so yes, this is replying to the two comments.

About the types, I now changed it to be exactly like the existing, wihch goes through double, and is shown to work on 32 bits.

Reply:
About the memcmp, I'll change that once those functions get merged in src/blockchain_db/lmdb

Reply:
> About the types, I now changed it to be exactly like the existing, wihch goes through double, and is shown to work on 32 bits.

I'm not sure what this means. I was originally commented on the:

```c++
uint64_t new_mapsize = mei.me_mapsize + bytes;
new_mapsize += (new_mapsize % mst.ms_psize);
```
and I wasn't immediately sure whether that could overflow before assignment.

Reply:
I mean I changed it to:
<pre>
  uint64_t new_mapsize = (double)mei.me_mapsize + bytes;
  new_mapsize += (new_mapsize % mst.ms_psize);
</pre>
which is what the existing code does in db_lmdb.cpp, and this one is shown to work on 32 bit archs since some people ran monerod there.

Reply:
Why `double` instead of `uint64_t`?

Reply:
i don't know. I don't think I wrote the original. I do agree uint64_t would seem more suited to it.

Reply:
Looking at git, it's from ND ages ago, so no asking what that was about. I think I'll just replace with uint64_t in the original separately, then change to uint64_t here in the patch.